fuse_publishers/src/graph_publisher.cpp

```diff
@@ -0,0 +1,287 @@
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <fuse_publishers/graph_publisher.h>
+
+#include <fuse_core/async_publisher.h>
+#include <fuse_core/graph.h>
+#include <fuse_core/transaction.h>
+#include <fuse_core/uuid.h>
+#include <pluginlib/class_list_macros.h>
+#include <ros/ros.h>
+#include <tf2/utils.h>
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+
+#include <exception>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+
+// Register this publisher with ROS as a plugin.
+PLUGINLIB_EXPORT_CLASS(fuse_publishers::GraphPublisher, fuse_core::Publisher);
+
+namespace fuse_publishers
+{
+
+GraphPublisher::GraphPublisher() :
+  fuse_core::AsyncPublisher(1)
+{
+}
+
+void GraphPublisher::onInit()
+{
+  // Read configuration from the parameter server
+  std::string device_str;
+  if (private_node_handle_.getParam("device_id", device_str))
+  {
+    device_id_ = fuse_core::uuid::from_string(device_str);
+  }
+  else if (private_node_handle_.getParam("device_name", device_str))
+  {
+    device_id_ = fuse_core::uuid::generate(device_str);
+  }
+  private_node_handle_.param("publish_optimized_graph", publish_optimized_graph_, publish_optimized_graph_);
+  private_node_handle_.param("queue_size", queue_size_, queue_size_);
+  private_node_handle_.param("frame_id", frame_id_, frame_id_);
+
+  // The graph doesn't have a frame ID
+  graph_msg_.header.frame_id = frame_id_;
+
+  // Advertise the topics
+  graph_publisher_ = private_node_handle_.advertise<fuse_msgs::Graph>("graph", queue_size_);
+}
+
+void GraphPublisher::onStart()
+{
+  graph_msg_.variables.clear();
+  graph_msg_.constraints.clear();
+
+  index_by_variable_uuid_.clear();
+  index_by_constraint_uuid_.clear();
+}
+
+void GraphPublisher::notifyCallback(
+  fuse_core::Transaction::ConstSharedPtr transaction,
+  fuse_core::Graph::ConstSharedPtr graph)
+{
+  updateGraphMessage(transaction);
+
+  if (graph_publisher_.getNumSubscribers() > 0)
+  {
+    // Unfortunately, the graph doesn't have a stamp and the transaction stamp is zero for merged transactions.
+    // Alternatively, we could use the most recent involved stamp in the transaction, but that wouldn't work if the
+    // transaction is empty
+    graph_msg_.header.stamp = ros::Time::now();
+
+    if (publish_optimized_graph_ && transaction->empty())
+    {
+      for (const auto& variable : graph->getVariables())
+      {
+        const auto variable_iter = index_by_variable_uuid_.find(variable.uuid());
+        if (variable_iter == index_by_variable_uuid_.end())
+        {
+          ROS_WARN_STREAM("Could not find variable " << fuse_core::uuid::to_string(variable.uuid())
```

Comment:
This and the other warning message below should never happen. Maybe I should remove them?!